memory_backend.go

```diff
@@ -0,0 +1,169 @@
+// their permissions
+type authBackend interface {
+	addAuth(*authorization) error
+	addMonitoringAuth(string) error
+	getAuthByID(id string) (authorization, error)
+	addSigner(signer.Signer)
+	getSigners() []signer.Signer
+	getSignerForUser(userID, signerID string) (signer.Signer, error)
+}
+
+// inMemoryBackend is an authBackend that loads a config and stores
+// that auth info in memory
+type inMemoryBackend struct {
+	auths       map[string]authorization
+	signerIndex map[string]int
+	signers     []signer.Signer
+}
+
+// newInMemoryAuthBackend returns an empty inMemoryBackend
+func newInMemoryAuthBackend() (backend *inMemoryBackend) {
+	return &inMemoryBackend{
+		auths:       make(map[string]authorization),
+		signerIndex: make(map[string]int),
+		signers:     []signer.Signer{},
+	}
+}
+
+// addAuth adds an authorization to the auth map or errors
+func (b *inMemoryBackend) addAuth(auth *authorization) (err error) {
+	_, getAuthErr := b.getAuthByID(auth.ID)
+	switch getAuthErr {
+	case nil:
+		return errors.Errorf("authorization id '%s' already defined, duplicates are not permitted", auth.ID)
+	case ErrAuthNotFound:
+		// this is what we want
+	default:
+		return errors.Wrapf(getAuthErr, "error finding auth with id '%s'", auth.ID)
+	}
+	if auth.HawkTimestampValidity != "" {
+		auth.hawkMaxTimestampSkew, err = time.ParseDuration(auth.HawkTimestampValidity)
+		if err != nil {
+			return err
+		}
+	} else {
+		auth.hawkMaxTimestampSkew = time.Minute
+	}
+	b.auths[auth.ID] = *auth
+	return b.addAuthToSignerIndex(auth)
+}
+
+// getAuthByID returns an authorization if it exists or nil. Call
+// addAuthorizations and addMonitoring first
+func (b *inMemoryBackend) getAuthByID(id string) (authorization, error) {
+	if auth, ok := b.auths[id]; ok {
+		return auth, nil
+	}
+	return authorization{}, ErrAuthNotFound
+}
+
+// addMonitoringAuth adds an authorization to enable the
+// tools/autograph-monitor
+func (b *inMemoryBackend) addMonitoringAuth(monitorKey string) error {
+	_, err := b.getAuthByID(monitorAuthID)
+	switch err {
+	case ErrAuthNotFound:
+	case nil:
+		return errors.Errorf("user 'monitor' is reserved for monitoring, duplication is not permitted")
+	default:
+		return errors.Errorf("error fetching 'monitor' auth: %q", err)
+	}
+	return b.addAuth(&authorization{
+		ID:                    monitorAuthID,
+		Key:                   monitorKey,
+		HawkTimestampValidity: "1m",
+		hawkMaxTimestampSkew:  time.Minute,
+	})
+}
+
+// getSignerId returns the signer identifier for the user. If a keyid
+// is specified, the corresponding signer is returned. If no signer is
+// found, an error is returned and the signer identifier is set to -1.
+func (b *inMemoryBackend) getSignerID(userid, keyid string) (int, error) {
+	tag := getSignerIndexTag(userid, keyid)
+	if _, ok := b.signerIndex[tag]; !ok {
+		if keyid == "" {
+			return -1, errors.Errorf("%q does not have a default signing key", userid)
```

Comment:
I wonder if this error is accurate. The default is just the first authorized signer for the user, so are we really saying that the user is not permitted to use any signer?

Reply:
and given line 139, it should be impossible to reach this code, right?

Reply:
This is from https://github.com/mozilla-services/autograph/blob/ebb4dffb0ae0ba89c6a79f58d225c5f920488205/authorize.go#L150-L159

> we really saying that the user is not permitted to use any signer?

Yes, and we said we didn't want any dangling users. I'm open to suggestions for how to reword it.

> and given line 139, it should be impossible to reach this code, right?

Yes, assuming auths aren't modified or reloaded since the app is loaded. This PR exists, because we said (in 1:1 or chat) we wanted to load auths on each request https://github.com/mozilla-services/autograph/pull/401#issuecomment-553929203